kdbx/src/main/java/org/linguafranca/pwdb/kdbx/KdbxKeyFile.java

```
@@ -45,33 +52,61 @@ public class KdbxKeyFile {
+                    Document doc = documentBuilder.parse(is);
+                    String version = (String) xpath.evaluate("//KeyFile/Meta/Version/text()", doc,
+                            XPathConstants.STRING);
+                    String data = (String) xpath.evaluate("//KeyFile/Key/Data/text()", doc, XPathConstants.STRING);
+                    if (data == null) {
                         return null;
                     }
+                    if (version.equals("2.0")) {
+
+                        byte[] hexData = Hex.decodeHex(data.replaceAll("\\s", ""));
+
+                        MessageDigest md = Encryption.getSha256MessageDigestInstance();
+                        byte[] computedHash = md.digest(hexData);
+
+                        String hashToCheck = (String) xpath.evaluate("//KeyFile/Key/Data/@Hash", doc,
+                                XPathConstants.STRING);
+                        byte[] verifiedHash = Hex.decodeHex(hashToCheck);
+
+                        boolean isHashVerified = Arrays.equals(Arrays.copyOf(computedHash, verifiedHash.length),
+                                                 verifiedHash);
+                        if (!isHashVerified) {
+                            throw new IllegalStateException("Hash mismatch error");
+                        }
+                        return hexData;
+                    }
+                    return Base64.decodeBase64(data.getBytes());
+                } else {
+                    //Any file compute the hash
+                    MessageDigest md = Encryption.getSha256MessageDigestInstance();
+                    byte[] keyFile = md.digest(inputBytes);
+                    return keyFile;
                 }
-                return hexData;
-                
             }
-            return Base64.decodeBase64(data.getBytes());
+
         } catch (Exception e) {
             throw new RuntimeException("Key File input stream cannot be null");
```

this catch all exception has too specific a cause attributed to it. I think it would be better to add IOException to method signature. Or if you really want to test for null then test for it, and throw an IllegalArgumentException